plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -86,6 +94,28 @@ data BuiltinRuntime val =
                                           -- out what it's going to cost.
         (ToCostingType n)
 
+instance NoThunks (BuiltinRuntime val) where
+    -- Skipping `_denot` in `allNoThunks` check, since it is supposed to be lazy
+    wNoThunks ctx (BuiltinRuntime sch _denot costing) =
+        allNoThunks
+            [ -- The order here is important: we must first check that `sch` doesn't have
+              -- thunks, before inspecting it in `noThunksInCosting`.
+              noThunks ctx sch
+            , noThunksInCosting ctx costing sch
+            ]
+
+    showTypeOf = const "PlutusCore.Builtin.Runtime.BuiltinRuntime"
+
+-- | Check whether the given `ToCostingType n` contains thunks.
+noThunksInCosting :: NoThunks.Context -> ToCostingType n -> RuntimeScheme n -> IO (Maybe ThunkInfo)
```

Comment:
This function never does anything to the costing argument, just recurses a bit and then calls `noThunks ctx costing`. So isn't it just equivalent to calling `noThunks ctx costing` directly?

Reply:
`ToCostingType n` is a type family so we can't directly call `noThunks ctx costing` because the type of `costing` depends on `n`.

Reply:
Can you `OnlyCheckWhnfNamed`-wrap the `ToCostingType` n argument and call `noThunks` on that instead of recursing?

Reply:
> Can you OnlyCheckWhnfNamed-wrap the ToCostingType n argument and call noThunks on that instead of recursing?

That means when `n ~ 'Z`, we are only checking that the `ExBudget` itself is not a thunk. But I think we also want to check that it doesn't contain nested thunks.

Reply:
> > 
> 
> That means when `n ~ 'Z`, we are only checking that the `ExBudget` itself is not a thunk. But I think we also want to check that it doesn't contain nested thunks.

Right, good point. I don't think we have any such cases, because builtins are not allowed to be unary and so there's always an argument, but that is not apparent in the types and it would be hacky to rely on this implicit assumption.

Could you document all of that?